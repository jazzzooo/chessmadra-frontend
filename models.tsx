import { RepertoireGrade, RepertoireMove } from "./utils/repertoire";

export interface LichessPuzzle {
  id: string;
  moves: string[];
  fen: string;
  popularity: number;
  tags: string[];
  gameLink: string;
  rating: number;
  ratingDeviation: number;
  numberPlays: number;
  allMoves: string[];
  maxPly: number;
}

export interface BlunderPuzzle {
  id: string;
  fen: string;
  bestMove: string;
  blunder: string;
  centipawnsLost: number;
}

export interface User {
  // paid?: string;
  email: string;
  // subscription: ApplicationSubscription;
  apiKey: string;
  betaAccess: boolean;
  eloRange: string;
}

export interface LichessGame {
  id: string;
  gameLink: string;
  result: number; // GameResult
  moves: string[];
  blackCentipawnLoss: number;
  whiteCentipawnLoss: number;
  numberMoves: number;
  whiteElo: number;
  blackElo: number;
  time: number;
  whiteBlunders: number;
  blackBlunders: number;
  whiteName: string;
  blackName: string;
  whiteMistakes: number;
  blackMistakes: number;
}

export interface RepertoireTemplate {
  title: string;
  id: string;
  followUp: string;
  line: string;
  num_moves: number;
  tags: string[];
}

export interface PlayerTemplate {
  id: string;
  meta: PlayerTemplateMeta;
  grade: RepertoireGrade;
}

export interface PlayerTemplateMeta {
  link: string;
  title: string;
  openings: string[];
  image: string;
  description: string;
}

export interface PositionReport {
  suggestedMoves: SuggestedMove[];
  stockfish?: StockfishReport;
  results: GameResultsDistribution;
  masterResults: GameResultsDistribution;
}

export interface StockfishReport {
  eval: number;
  mate: number;
  nodesK: number;
}

export interface SuggestedMove {
  sanPlus: string;
  stockfish?: StockfishReport;
  results: GameResultsDistribution;
  masterResults: GameResultsDistribution;
  percentagePlayedAtLevel: number;
}

export interface GameResultsDistribution {
  white: number;
  black: number;
  draw: number;
}
