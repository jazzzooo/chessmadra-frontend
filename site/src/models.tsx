export interface LichessPuzzle {
  id: string;
  moves: string[];
  fen: string,
  popularity: number,
  tags: string[],
  gameLink: string,
  rating: number,
  ratingDeviation: number,
  numberPlays: number,
  allMoves: string[],
  maxPly: number,
}

