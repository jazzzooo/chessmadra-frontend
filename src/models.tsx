import { RepertoireGrade } from "./utils/repertoire";

export interface LichessPuzzle {
  id: string;
  moves: string[];
  fen: string;
  popularity: number;
  tags: string[];
  gameLink: string;
  rating: number;
  ratingDeviation: number;
  numberPlays: number;
  allMoves: string[];
  maxPly: number;
}

export interface BlunderPuzzle {
  id: string;
  fen: string;
  bestMove: string;
  blunder: string;
  centipawnsLost: number;
}

export interface User {
  id: string;
  // paid?: string;
  email: string;
  // subscription: ApplicationSubscription;
  apiKey: string;
  betaAccess: boolean;
  // The behind-the-scenes range
  eloRange: string;
  temporary: boolean;
  isAdmin: boolean;
  missThreshold: number;
  // The range/system that's visible to the user
  ratingRange: string;
  ratingSystem: string;
}

export interface LichessGame {
  id: string;
  gameLink: string;
  result: number; // GameResult
  moves: string[];
  blackCentipawnLoss: number;
  whiteCentipawnLoss: number;
  numberMoves: number;
  whiteElo: number;
  blackElo: number;
  time: number;
  whiteBlunders: number;
  blackBlunders: number;
  whiteName: string;
  blackName: string;
  whiteMistakes: number;
  blackMistakes: number;
}

export interface RepertoireTemplate {
  title: string;
  id: string;
  followUp: string;
  line: string;
  num_moves: number;
  tags: string[];
}

export interface PlayerTemplate {
  id: string;
  meta: PlayerTemplateMeta;
  grade: RepertoireGrade;
}

export interface PlayerTemplateMeta {
  link: string;
  title: string;
  openings: string[];
  image: string;
  description: string;
}

export interface PositionReport {
  epd: string;
  suggestedMoves: SuggestedMove[];
  pawnStructure: PawnStructure;
  stockfish?: StockfishReport;
  results: GameResultsDistribution;
  masterResults: GameResultsDistribution;
}

export interface PawnStructure {
  name: string;
  reversed: boolean;
}

export interface StockfishReport {
  eval: number;
  mate: number;
  nodesK: number;
}

export interface SuggestedMove {
  sanPlus: string;
  epdAfter: string;
  stockfish?: StockfishReport;
  results: GameResultsDistribution;
  masterResults: GameResultsDistribution;
  percentagePlayedAtLevel: number;
  annotation: string;
}

export interface GameResultsDistribution {
  white: number;
  black: number;
  draw: number;
}

export interface EcoCode {
  epd: string;
  fullName: string;
  code: string;
}

export interface PawnStructureDetails {
  id: string;
  name: string;
  moreInfoLink: string;
  plans: string;
  pawnEpds: string;
  opponentPlans: string;
}

export interface MoveAnnotationReview {
  epd: string;
  san: string;
  annotations: {
    userId: string;
    text: string;
  }[];
}

export interface MoveAnnotation {
  epd: string;
  san: string;
  text: string;
  userId: string;
}
