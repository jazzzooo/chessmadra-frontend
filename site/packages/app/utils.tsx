export const times = (x) => (f) => {
  let results = [];
  for (let i = 0; i < x; i++) {
    results.push(f(i));
  }
  return results;
};
