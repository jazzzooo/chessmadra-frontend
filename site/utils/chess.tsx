import { COLUMNS, ROWS } from "../types/Chess";
import { indexOf } from "lodash";
import { Chess, Move } from "@lubert/chess.ts";

export const getSquareOffset = (square: string, flipped: boolean) => {
  const [file, rank] = square.split("");
  let x = indexOf(COLUMNS, file);
  let y = 7 - indexOf(ROWS, parseInt(rank));
  if (flipped) {
    x = 7 - x;
    y = 7 - y;
  }
  return { x: x / 8, y: y / 8 };
};

export const isCheckmate = (move: Move, position: Chess) => {
  position.move(move);
  let isCheckMate = position.inCheckmate();
  position.undo();
  return isCheckMate;
};
