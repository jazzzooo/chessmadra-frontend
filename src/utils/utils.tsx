export const times = (x) => (f) => {
  const results = [];
  for (let i = 0; i < x; i++) {
    results.push(f(i));
  }
  return results;
};

export const compareFloats = (a, b) => {
  return Math.abs(a - b) < Number.EPSILON;
};
