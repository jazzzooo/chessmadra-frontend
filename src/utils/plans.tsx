import { Chess } from "@lubert/chess.ts";
import { Square } from "@lubert/chess.ts/dist/types";
import { Plan } from "app/models";
import { sortBy, filter, uniqBy, take, cloneDeep } from "lodash-es";
import { Side, toSide } from "./repertoire";

export const getTopPlans = (
  _plans: Plan[],
  side: Side,
  board: Chess
): Plan[] => {
  let plans = _plans;
  plans = filter(plans, (p) => p.side === side);

  // @ts-ignore
  let byFromSquare: Record<Square, Plan[]> = {};
  plans.forEach((p) => {
    if (byFromSquare[p.fromSquare]) {
      byFromSquare[p.fromSquare].push(p);
    } else {
      byFromSquare[p.fromSquare] = [p];
    }
  });
  let nonGhostPlans: Plan[] = [];
  type SquareMove = string;
  let recurse = (plan: Plan, seenMoves: Set<SquareMove>) => {
    nonGhostPlans.push(plan);
    byFromSquare[plan.fromSquare].forEach((p) => {
      if (seenMoves.has(p.fromSquare + p.toSquare)) {
        seenMoves.add(p.fromSquare + p.toSquare);
        recurse(p, cloneDeep(seenMoves));
      }
    });
  };
  plans.forEach((plan) => {
    let piece = board.get(plan.fromSquare);
    if (piece && toSide(piece.color) === side) {
      byFromSquare[plan.fromSquare].forEach((p) => {
        recurse(p, new Set());
      });
    }
  });

  // plans = nonGhostPlans;
  plans = uniqBy(plans, (p) => `${p.san}-${p.toSquare}-${p.toSquare}`);
  plans = sortBy(plans, (p) => -p.occurences);

  if (plans.length <= 1) {
    return [];
  }
  return take(plans, 7);
};
