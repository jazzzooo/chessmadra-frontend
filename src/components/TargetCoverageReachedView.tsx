import React, { useState } from "react";
import { Spacer } from "app/Space";
import {
  capitalize,
  filter,
  isEmpty,
  some,
  sortBy,
  take,
  find,
  every,
} from "lodash-es";
import { useResponsive } from "app/utils/useResponsive";
import { CMTextInput } from "./TextInput";
import { SidebarTemplate } from "./SidebarTemplate";
import client from "app/client";
import {
  quick,
  useBrowsingState,
  useSidebarState,
  useUserState,
} from "app/utils/app_state";
import { useRepertoireState } from "app/utils/app_state";
import { CMText } from "./CMText";
import { s, c } from "app/styles";
import { intersperse } from "app/utils/intersperse";
import { Plan } from "app/models";
import { Side } from "app/utils/repertoire";
import { View } from "react-native";
import { getTopPlans, MetaPlan } from "app/utils/plans";
import { useHovering } from "app/hooks/useHovering";

type PlanSection = JSX.Element | JSX.Element[] | string;

export const TargetCoverageReachedView = () => {
  const responsive = useResponsive();
  const user = useUserState((s) => s.user);
  const [feedback, setFeedback] = useState("");
  const [email, setEmail] = useState("");
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState(false);
  const submitFeedback = () => {
    if (isEmpty(feedback)) {
      return;
    }
    setLoading(true);
    client
      .post("/api/v1/submit-feedback", {
        feedback,
        email: user?.email ?? email ?? null,
      })
      .then(() => {
        setSuccess(true);
        setLoading(false);
      });
  };
  let [currentEpd] = useSidebarState(([s]) => [s.currentEpd]);
  const [activeSide] = useBrowsingState(([s, rs]) => [s.activeSide]);
  let [plans] = useRepertoireState((s) => [
    getTopPlans(
      s.positionReports[currentEpd]?.plans ?? [],
      s.browsingState.activeSide,
      s.browsingState.chessboardState.position
    ),
  ]);

  let planSections = [];
  planSections.push(getCastlingPlanSection(plans, activeSide));
  planSections.push(getPawnPlansSection(plans, activeSide));
  planSections.push(getPiecePlansSection(plans, activeSide));
  planSections = planSections.filter((p) => p);
  return (
    <SidebarTemplate
      header={"You've reached your target depth!  âœ…"}
      actions={[
        {
          onPress: () => {
            quick((s) => {
              s.repertoireState.browsingState.addPendingLine();
            });
          },
          style: "primary",
          text: "I'm done, save this line to my repertoire",
        },
        {
          onPress: () => {
            quick((s) => {
              s.repertoireState.browsingState.moveSidebarState("right");
              s.repertoireState.browsingState.sidebarState.targetCoverageReachedState.visible =
                false;
            });
          },
          style: "primary",
          text: "Keep adding moves to this line",
        },
      ]}
      loading={loading && "Submitting feedback..."}
      bodyPadding={true}
    >
      {/*<CMText style={s(c.sidebarDescriptionStyles(responsive))}>
        Feature request? Bug report? Etc? Let us know
      </CMText>
      <Spacer height={12} />
*/}
      <Spacer height={12} />
      <CMText
        style={s(c.weightBold, c.fontSize(14), c.fg(c.colors.textPrimary))}
      >
        How to play from here
      </CMText>
      <Spacer height={12} />
      <View>
        {intersperse(
          planSections.map((section, i) => {
            return (
              <View style={s()} key={i}>
                <CMText style={s()}>{section}</CMText>
              </View>
            );
          }),
          (k) => {
            return <Spacer key={k} height={12} />;
          }
        )}
      </View>
    </SidebarTemplate>
  );
};

// TODO: move text elements should have a component that has hover behavior that
// highlights the plan on the board

function getCastlingPlanSection(plans: MetaPlan[], side: Side): PlanSection {
  let queenside = find(plans, (p) => p.plan.san === "O-O-O");
  let kingside = find(plans, (p) => p.plan.san === "O-O");
  if (!(queenside || kingside)) {
    return null;
  } else if (queenside && kingside) {
    let queensideMoreCommon =
      queenside.plan.occurences > kingside.plan.occurences;
    return (
      <>
        You can castle to either side, although{" "}
        <PlanMoveText plan={queensideMoreCommon ? queenside : kingside}>
          castling {queensideMoreCommon ? "queenside" : "kingside"}
        </PlanMoveText>{" "}
        is most common among experts
      </>
    );
  } else if (kingside) {
    return (
      <>
        <PlanMoveText plan={kingside}>Castling kingside</PlanMoveText> is best
      </>
    );
  } else if (queenside) {
    return (
      <>
        {capitalize(side)} tends to{" "}
        <PlanMoveText plan={queenside}>Castling queenside</PlanMoveText> is best
      </>
    );
  }
}

function getPawnPlansSection(plans: MetaPlan[], side: Side): PlanSection {
  let pawnPlans = filter(plans, (p) =>
    some(["a", "b", "c", "d", "e", "f", "g", "h"], (f) =>
      p.plan.san?.startsWith(f)
    )
  );
  if (isEmpty(pawnPlans)) {
    return null;
  }
  return (
    <>
      <PlanMoves plans={pawnPlans.map((p) => p.plan)} />{" "}
      {pawnPlans.length > 1 ? "are common pawn moves" : "is a common pawn move"}
    </>
  );
}

function getPiecePlansSection(plans: MetaPlan[], side: Side): PlanSection {
  let piecePlans = filter(plans, (p) =>
    some(["N", "B", "R", "Q", "K"], (f) => p.plan.san?.startsWith(f))
  );
  if (isEmpty(piecePlans)) {
    return null;
  }
  return (
    <>
      Common piece moves include{" "}
      <PlanMoves plans={piecePlans.map((p) => p.plan)} />
    </>
  );
}

const PlanMoveText = ({ plan, children }) => {
  const { hovering, hoveringProps } = useHovering(
    () => {
      quick((s) => {
        s.repertoireState.browsingState.chessboardState.focusedPlan = plan;
      });
    },
    () => {
      quick((s) => {
        s.repertoireState.browsingState.chessboardState.focusedPlan = null;
      });
    }
  );
  return (
    <View style={s(c.inlineBlock, c.clickable)} {...hoveringProps}>
      <CMText style={s(c.weightBold)}>{children}</CMText>
    </View>
  );
};

const PlanMoves = ({ plans }: { plans: Plan[] }) => {
  return (
    <CMText style={s()}>
      {intersperse(
        plans.map((plan, i) => {
          return <PlanMoveText plan={plan}>{plan.san}</PlanMoveText>;
        }),
        (k, isLast) => {
          return (
            <CMText key={k} style={s()}>
              {isLast ? (plans.length > 2 ? ", and " : " and ") : ", "}
            </CMText>
          );
        }
      )}
    </CMText>
  );
};
