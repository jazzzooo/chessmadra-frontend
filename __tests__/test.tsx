import { describe, expect, test } from "@jest/globals";
import { Chess } from "@lubert/chess.ts";
import { getPlanSections } from "app/components/TargetCoverageReachedView";
import { Plan } from "app/models";
import { getTopPlans } from "app/utils/plans";

describe("plan describer", () => {
  test("Basics", () => {
    let plans = [
      {
        san: "O-O-O",
        occurences: 1,
        fromSquare: "e1",
        toSquare: "c1",
        side: "white",
      } as Plan,
      {
        san: "c3",
        occurences: 1,
        fromSquare: "c2",
        toSquare: "c3",
        side: "white",
      } as Plan,
      {
        san: "d3",
        occurences: 1,
        fromSquare: "d2",
        toSquare: "d3",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Knight development", () => {
    let plans = [
      {
        san: "Nf3",
        occurences: 1,
        fromSquare: "g1",
        toSquare: "f3",
        side: "white",
      } as Plan,
      {
        san: "Nf3",
        occurences: 1,
        fromSquare: "g1",
        toSquare: "e2",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Bishop development", () => {
    let plans = [
      {
        san: "Bf4",
        occurences: 1,
        fromSquare: "c1",
        toSquare: "f4",
        side: "white",
      } as Plan,
      {
        san: "Bc4",
        occurences: 1,
        fromSquare: "f1",
        toSquare: "c4",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Bishop development", () => {
    let plans = [
      {
        san: "Bf4",
        occurences: 1,
        fromSquare: "c1",
        toSquare: "f4",
        side: "white",
      } as Plan,
      {
        san: "Bc4",
        occurences: 1,
        fromSquare: "f1",
        toSquare: "c4",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Bishop movement", () => {
    let plans = [
      {
        san: "Bf4",
        occurences: 1,
        fromSquare: "d2",
        toSquare: "f4",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Bishop movement", () => {
    let plans = [
      {
        san: "Bf4",
        occurences: 1,
        fromSquare: "d2",
        toSquare: "f4",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Pawn pushes", () => {
    let plans = [
      {
        san: "c3",
        occurences: 1,
        fromSquare: "c2",
        toSquare: "c3",
        side: "white",
      } as Plan,
      {
        san: "c4",
        occurences: 1,
        fromSquare: "c3",
        toSquare: "c4",
        side: "white",
      } as Plan,
    ];
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
  test("Poison pawn", () => {
    let plans = [
      {
        fromSquare: "d1",
        toSquare: "d2",
        san: "Qd2",
        side: "white",
        occurences: 3315,
      },
      {
        fromSquare: "a1",
        toSquare: "b1",
        san: "Rb1",
        side: "white",
        occurences: 2288,
      },
      {
        fromSquare: "f1",
        toSquare: "e2",
        san: "Be2",
        side: "white",
        occurences: 2094,
      },
      {
        fromSquare: "g5",
        toSquare: "f6",
        san: "Bxf6",
        side: "white",
        occurences: 1936,
      },
      {
        fromSquare: "d4",
        toSquare: "b3",
        san: "Nb3",
        side: "white",
        occurences: 1879,
      },
      {
        fromSquare: "e1",
        toSquare: "a1",
        san: "O-O-O",
        side: "white",
        occurences: 1797,
      },
      {
        fromSquare: "e4",
        toSquare: "e5",
        san: "e5",
        side: "white",
        occurences: 1613,
      },
      {
        fromSquare: "f4",
        toSquare: "f5",
        san: "f5",
        side: "white",
        occurences: 1604,
      },
      {
        fromSquare: "g5",
        toSquare: "h4",
        san: "Bh4",
        side: "white",
        occurences: 1283,
      },
      {
        fromSquare: "d4",
        toSquare: "c6",
        san: "Nxc6",
        side: "white",
        occurences: 1214,
      },
      {
        fromSquare: "b8",
        toSquare: "c6",
        san: "Nc6",
        side: "black",
        occurences: 2711,
      },
      {
        fromSquare: "b6",
        toSquare: "b2",
        san: "Qxb2",
        side: "black",
        occurences: 2542,
      },
      {
        fromSquare: "b2",
        toSquare: "a3",
        san: "Qa3",
        side: "black",
        occurences: 2424,
      },
      {
        fromSquare: "f8",
        toSquare: "e7",
        san: "Be7",
        side: "black",
        occurences: 2275,
      },
      {
        fromSquare: "h7",
        toSquare: "h6",
        san: "h6",
        side: "black",
        occurences: 1607,
      },
      {
        fromSquare: "g7",
        toSquare: "f6",
        san: "gxf6",
        side: "black",
        occurences: 1521,
      },
      {
        fromSquare: "d6",
        toSquare: "e5",
        san: "dxe5",
        side: "black",
        occurences: 1277,
      },
      {
        fromSquare: "b7",
        toSquare: "c6",
        san: "bxc6",
        side: "black",
        occurences: 1110,
      },
      {
        fromSquare: "b8",
        toSquare: "d7",
        san: "Nbd7",
        side: "black",
        occurences: 1005,
      },
      {
        fromSquare: "f7",
        toSquare: "e6",
        san: "fxe6",
        side: "black",
        occurences: 872,
      },
    ];
    // @ts-ignore
    expect(getPlanSections(plans, "white", new Chess())).toMatchSnapshot();
  });
});
