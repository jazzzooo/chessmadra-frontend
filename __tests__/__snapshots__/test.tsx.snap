// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plan describer Basics 1`] = `
Array [
  <React.Fragment>
    White
     typically
     
    <PlanMoveText
      plan={
        Object {
          "directionChanged": false,
          "id": "e1 c1 k",
          "piece": "k",
          "plan": Object {
            "fromSquare": "e1",
            "occurences": 1,
            "san": "O-O-O",
            "side": "white",
            "toSquare": "c1",
          },
          "subsequentMove": false,
        }
      }
    >
      castles queenside
    </PlanMoveText>
  </React.Fragment>,
  <React.Fragment>
    Common pawn moves are
     
    <PlanMoves
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "c2 c3 p",
            "piece": "p",
            "plan": Object {
              "fromSquare": "c2",
              "occurences": 1,
              "san": "c3",
              "side": "white",
              "toSquare": "c3",
            },
            "subsequentMove": false,
          },
          Object {
            "directionChanged": false,
            "id": "d2 d3 p",
            "piece": "p",
            "plan": Object {
              "fromSquare": "d2",
              "occurences": 1,
              "san": "d3",
              "side": "white",
              "toSquare": "d3",
            },
            "subsequentMove": false,
          },
        ]
      }
    />
     
  </React.Fragment>,
]
`;

exports[`plan describer Bishop development 1`] = `
Array [
  <React.Fragment>
    The 
    dark-squared bishop
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "c1 f4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "c1",
              "occurences": 1,
              "san": "Bf4",
              "side": "white",
              "toSquare": "f4",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
  <React.Fragment>
    The 
    light-squared bishop
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "f1 c4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "f1",
              "occurences": 1,
              "san": "Bc4",
              "side": "white",
              "toSquare": "c4",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
]
`;

exports[`plan describer Bishop development 2`] = `
Array [
  <React.Fragment>
    The 
    dark-squared bishop
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "c1 f4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "c1",
              "occurences": 1,
              "san": "Bf4",
              "side": "white",
              "toSquare": "f4",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
  <React.Fragment>
    The 
    light-squared bishop
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "f1 c4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "f1",
              "occurences": 1,
              "san": "Bc4",
              "side": "white",
              "toSquare": "c4",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
]
`;

exports[`plan describer Bishop movement 1`] = `
Array [
  <React.Fragment>
    Common piece moves include 
    <PlanMoves
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "d2 f4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "d2",
              "occurences": 1,
              "san": "Bf4",
              "side": "white",
              "toSquare": "f4",
            },
            "subsequentMove": false,
          },
        ]
      }
    />
  </React.Fragment>,
]
`;

exports[`plan describer Bishop movement 2`] = `
Array [
  <React.Fragment>
    Common piece moves include 
    <PlanMoves
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "d2 f4 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "d2",
              "occurences": 1,
              "san": "Bf4",
              "side": "white",
              "toSquare": "f4",
            },
            "subsequentMove": false,
          },
        ]
      }
    />
  </React.Fragment>,
]
`;

exports[`plan describer Knight development 1`] = `
Array [
  <React.Fragment>
    The 
    kingside knight
     
    can be developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "g1 f3 n",
            "piece": "n",
            "plan": Object {
              "fromSquare": "g1",
              "occurences": 1,
              "san": "Nf3",
              "side": "white",
              "toSquare": "f3",
            },
            "subsequentMove": false,
          },
          Object {
            "directionChanged": false,
            "id": "g1 e2 n",
            "piece": "n",
            "plan": Object {
              "fromSquare": "g1",
              "occurences": 1,
              "san": "Nf3",
              "side": "white",
              "toSquare": "e2",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
]
`;

exports[`plan describer Pawn pushes 1`] = `
Array [
  <React.Fragment>
    A common pawn move is
     
    <PlanMoves
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "c2 c3 p",
            "piece": "p",
            "plan": Object {
              "fromSquare": "c2",
              "occurences": 1,
              "san": "c3",
              "side": "white",
              "toSquare": "c3",
            },
            "subsequentMove": false,
          },
        ]
      }
    />
     
  </React.Fragment>,
]
`;

exports[`plan describer Poison pawn 1`] = `
Array [
  <React.Fragment>
    White
     typically
     
    <PlanMoveText
      plan={
        Object {
          "directionChanged": false,
          "id": "e1 a1 k",
          "piece": "k",
          "plan": Object {
            "fromSquare": "e1",
            "occurences": 1797,
            "san": "O-O-O",
            "side": "white",
            "toSquare": "a1",
          },
          "subsequentMove": false,
        }
      }
    >
      castles queenside
    </PlanMoveText>
  </React.Fragment>,
  <React.Fragment>
    The 
    queen
     
    belongs on
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "d1 d2 q",
            "piece": "q",
            "plan": Object {
              "fromSquare": "d1",
              "occurences": 3315,
              "san": "Qd2",
              "side": "white",
              "toSquare": "d2",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
  <React.Fragment>
    The 
    queenside rook
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "a1 b1 r",
            "piece": "r",
            "plan": Object {
              "fromSquare": "a1",
              "occurences": 2288,
              "san": "Rb1",
              "side": "white",
              "toSquare": "b1",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
  <React.Fragment>
    The 
    light-squared bishop
     
    is generally developed to
     
    <PlanMoves
      exclusive={true}
      metaPlans={
        Array [
          Object {
            "directionChanged": false,
            "id": "f1 e2 b",
            "piece": "b",
            "plan": Object {
              "fromSquare": "f1",
              "occurences": 2094,
              "san": "Be2",
              "side": "white",
              "toSquare": "e2",
            },
            "subsequentMove": false,
          },
        ]
      }
      stripPieceSymbol={true}
    />
  </React.Fragment>,
]
`;
